src/Microsoft.OData.Core/UriParser/Parsers/SelectExpandTermParser.cs

```diff
@@ -109,25 +109,38 @@ private void CheckPathLength(int pathLength)
         /// <returns>A parsed PathSegmentToken representing the next segment in this path.</returns>
         private PathSegmentToken ParseSegment(PathSegmentToken previousSegment, bool allowRef)
         {
-            // TODO $count is defined in specification for expand, it is not supported now. Also note $count is not supported with star as expand option.
-            if (this.lexer.CurrentToken.Text.StartsWith("$", StringComparison.Ordinal) && (!allowRef || this.lexer.CurrentToken.Text != UriQueryConstants.RefSegment))
+            if (this.lexer.CurrentToken.Text.StartsWith("$", StringComparison.Ordinal)
+                && (!allowRef || this.lexer.CurrentToken.Text != UriQueryConstants.RefSegment)
+                && this.lexer.CurrentToken.Text != UriQueryConstants.CountSegment)
             {
                 throw new ODataException(ODataErrorStrings.UriSelectParser_SystemTokenInSelectExpand(this.lexer.CurrentToken.Text, this.lexer.ExpressionText));
             }
 
-            // Some check here to throw exception, both prop1/*/prop2 and */$ref/prop will throw exception, both are for $expand cases
+            // Some check here to throw exception, prop1/*/prop2 and */$ref/prop and prop1/$count/prop2 will throw exception, all are $expand cases.
             if (!isSelect)
             {
                 if (previousSegment != null && previousSegment.Identifier == UriQueryConstants.Star && this.lexer.CurrentToken.GetIdentifier() != UriQueryConstants.RefSegment)
                 {
-                    // Star can only be followed with $ref
+                    // Star can only be followed with $ref. $count is not supported with star as expand option
                     throw new ODataException(ODataErrorStrings.ExpressionToken_OnlyRefAllowWithStarInExpand);
                 }
                 else if (previousSegment != null && previousSegment.Identifier == UriQueryConstants.RefSegment)
                 {
                     // $ref should not have more property followed.
                     throw new ODataException(ODataErrorStrings.ExpressionToken_NoPropAllowedAfterRef);
                 }
+                else if (previousSegment != null && previousSegment.Identifier == UriQueryConstants.CountSegment)
+                {
+                    // $count should not have more property followed. e.g $expand=NavProperty/$count/MyProperty
+                    throw new ODataException(ODataErrorStrings.ExpressionToken_NoPropAllowedAfterDollarCount);
+                }
+            }
+
+            if (this.lexer.CurrentToken.Text == UriQueryConstants.CountSegment && isSelect)
+            {
+                // $count in only allowed in $expand. e.g $expand=NavProperty/$count
+                // It is not allowed in $select e.g $select=NavProperty/$count
+                throw new ODataException(ODataErrorStrings.ExpressionToken_DollarCountOnlyAllowedInExpand);

```

Comment:
The error message could be confusing, because $count can use in $filter, $orderby, etc.
If that's not allowed in $select, could be change it as "NotAllowedInSelect"?

Reply:
Fixed!